Support Kubernetes Job to build Docker image

hdt91 opened the pull request:
This PR add the basic support to create a job on a K8s cluster running a predefined image. This image contains a script to clone the specified repository, commit, tag, etc. passed as environment variables, then optionally push the newly built image and reports the result log.

/cc @jonmoter @grosser 

### Tasks
- Fix CodeClimate and code style violations
- Create the build job image and push to registry
- Test the whole flow on the staging cluster
- Fix any issue after code review ?

### References
 - Jira link: https://zendesk.atlassian.net/browse/PAAS-141

### Risks
- Level: Low

jonmoter commented on app/models/docker_builder_service.rb:
```diff
@@ -21,9 +22,35 @@ def run!(image_name: nil, push: false, tag_as_latest: false)
+          output.puts "### Failed to get the image digest"
+        end
+
+        build.save!
```
Refactor this logic into a separate method, since `run!` is getting pretty big.

jonmoter commented on plugins/kubernetes/app/models/kubernetes/job_executor.rb:
```diff
@@ -0,0 +1,96 @@
+# Execute a job that can build/push Docker images
+# and write progress to the local job output.
+module Kubernetes
+  class JobExecutor
```
Call this something like `BuildJobExecutor`, since it's specifically for executing a job that builds Docker images, rather something that executes jobs generically.

jonmoter commented on plugins/kubernetes/app/models/kubernetes/job_executor.rb:
```diff
@@ -0,0 +1,96 @@
+    end
+
+    def build!(reference, repo:, tag:, push: false, registry:)
+      job_log = ""
```
Seems strange to have this write to `@output`, and also to return a string with the job output.  Why not just write all output to `@output`?

hdt91 replied:
The intention is to return the specific job log to the caller so it can parse only the necessary information.

jonmoter commented on plugins/kubernetes/app/models/kubernetes/job_executor.rb:
```diff
@@ -0,0 +1,96 @@
+      k8s_job.first_container.args = [@job.project.repository_url, reference, repo, tag, push ? "yes" : "no"]
+      k8s_job.first_container.env = [
+        {name: 'DOCKER_REGISTRY_USER', value: registry[:username] || ""},
+        {name: 'DOCKER_REGISTRY_PASS', value: registry[:password] || ""},
```
Not a good practice to put sensitive information into the ENV of an object. Should really use a secret, or put something into Vault.

Also, Kubernetes has its own mechanism to authenticate Docker with remote registries. We can probably leverage that. http://kubernetes.io/docs/user-guide/images

hdt91 replied:
I can remove this since it is optional. The authentication happens at the job container so just need to change the image. Is there any documentation about the recommended way that we use auth/secret in the cluster, such as just assume the cluster already has the secret object?

jonmoter commented on plugins/kubernetes/app/models/kubernetes/job_executor.rb:
```diff
@@ -0,0 +1,96 @@
+
+      contents = File.read(build_job_config_path)
+      k8s_job = Kubernetes::JobConfigFile.new(contents, build_job_config_path).job
+      k8s_job.metadata.name = "docker-build-job-#{SecureRandom.hex(16)}"
```
Might be nice to name the job based on the project and build id.  Like "docker-build-help-center-12421", so that it's more obvious what the job is for.

Also, add some labels like the name of the project, and something labeling it as a docker build job, so we can easily query on these kind of jobs.

hdt91 replied:
Makes sense and a nice addition. Is there any guideline about the convention to attach labels to different Kubernetes resources?

jonmoter commented on plugins/kubernetes/app/models/kubernetes/job_executor.rb:
```diff
@@ -0,0 +1,96 @@
+
+      @output.puts "### Running a remote Docker build job: #{k8s_job.metadata.name}"
+      extension_client.create_job(k8s_job)
+      success, job_obj = wait_for_job_status(k8s_job)
```
Enhancement: use `watch_pod_log` and write the output to the `@output` buffer, so that the end-user gets feedback before the job completes.

hdt91 commented:
There is another pending PR that actually is sort of a superset for this one. What is your opinion on this?
https://github.com/zendesk/samson/pull/1061

jonmoter commented:
I think that the other PR has some code that is more general-purpose for using Kubernetes jobs.  However, there's also a lot of copy-pasted code there that needs cleanup.

So my hunch is that we'd be best off cleaning up and merging the other PR, and then adapting your code to fit in with that.

Btw, I'm sorry it's taken so long to review and merge your code here.  I've been really heads-down trying to get the cluster running in our staging environments, wrangling chef cookbooks, etc.  So without a stable Kubernetes cluster, it's not useful to merge in code that relies on the cluster for building. (In general, this project has required a lot more DevOps work than Dev work, for better or worse.)

hdt91 commented:
No worries, that PR actually covers a few TODO that I wanted originally but decided not to avoid spilling the story scope and explore where a job might fit in. I would prefer to have that one get merged nicely, since the adoption should be quite smooth based on what I have read. When everything is good, I will clean up this PR and ping you guys again :D

grosser commented:
FYI job support is ready in #1092 will hopefully be merged shorty, not sure how helpful that will be though ...

hdt91 commented:
I have cleaned up the PR based on your recent commits and #1092. Please take a look. Btw how to view the Code Climate result?

grosser commented on plugins/kubernetes/app/models/kubernetes/config_file.rb:
```diff
@@ -0,0 +1,33 @@
+module Kubernetes
```
this should not be necessary ...

grosser commented on plugins/kubernetes/app/models/kubernetes/config_verifier.rb:
```diff
@@ -0,0 +1,112 @@
+module Kubernetes
+  class ConfigVerifier
```
already have this too ...

grosser commented:
why do we need multiple verifiers 1 should be able to handle all resource types ...

grosser commented on plugins/kubernetes/app/models/kubernetes/build_job_executor.rb:
```diff
@@ -0,0 +1,133 @@
+
+    def fill_job_details(k8s_job, build, project, tag:, push: false, tag_as_latest: false)
+      # Fill in some information to easily query the job resource
+      project_name = project.name.downcase.tr('_', '-')
```
use `.permalink` ?

grosser commented on plugins/kubernetes/app/models/kubernetes/build_job_executor.rb:
```diff
@@ -0,0 +1,133 @@
+      # Fill in some information to easily query the job resource
+      project_name = project.name.downcase.tr('_', '-')
+      labels = { project: project_name, role: "docker-build-job" }
+      k8s_job[:metadata][:name] = "docker-build-#{project_name}-#{SecureRandom.hex(7)}"
```
put the build id in there too ?

grosser commented on plugins/kubernetes/app/models/kubernetes/build_job_executor.rb:
```diff
@@ -0,0 +1,133 @@
+      k8s_job[:spec][:template][:metadata][:labels].update(labels)
+
+      # Necessary to query the pod running the job later
+      k8s_job[:spec][:template][:metadata][:labels][:jobName] = k8s_job[:metadata][:name]
```
I think so far we used underscopre labels -> job_name

grosser commented on plugins/kubernetes/app/models/kubernetes/build_job_executor.rb:
```diff
@@ -0,0 +1,133 @@
+        env: [{name: 'DOCKER_REGISTRY', value: @registry[:serveraddress] }],
+        args: [
+          project.repository_url, build.git_sha, project.docker_repo,
+          tag, push ? "yes" : "no", tag_as_latest ? "yes" : "no"
```
this looks a bit crazy ... maybe parentheses can help  tag, (push ? ...), (tag_as_latest ? ...)

grosser commented on plugins/kubernetes/app/models/kubernetes/build_job_executor.rb:
```diff
@@ -0,0 +1,133 @@
+          # a new watch stream.
+          if pod_name != pod.name
+            pod_name = pod.name
+            client.watch_pod_log(pod.name, pod.namespace).each do |line|
```
I like the streaming part, should give instant feedback

grosser commented on plugins/kubernetes/app/models/kubernetes/build_job_executor.rb:
```diff
@@ -0,0 +1,133 @@
+        sleep TICK # Sleep a bit to wait for the server
+        begin
+          job = Kubernetes::Api::Job.new(
+            extension_client.get_job(k8s_job[:metadata][:name], k8s_job[:metadata][:namespace])
```
maybe extract `job_name` it's used 3 times

grosser commented on plugins/kubernetes/config/build_job.yml:
```diff
@@ -0,0 +1,34 @@
+---
```
could add a top-level comment here explaining what it does

grosser commented:
that looks reasonable ...

things I'd like to change:
 - extract as much view logic as possible into new plugin hooks
 - ideally reuse existing verifier/config loader
 - ideally extract remove_job logic into kubernetes ... but that might be tricky ...
 - maybe rename remote_job attribute -> kubernetes_job for clarity

hdt91 commented:
Thanks for all those suggestions. May I know more about what is your thought on the third point?

grosser commented:
kubernetes related things should live in the kubernetes plugin 
 - for simpler grepping/overview
 - to not distract people working on 'regular' samson
 - to be able to remove it if kubernetes ever goes out of fashion
 - to be able to use the same hooks to plug in another container framework

hdt91 commented:
I have reused the config reader/verifier, cleaned up the variable name and moved the view parts into kubernetes plugin directory.

It is indeed a bit tricky to move the part where the build job execution is done in the docker image service as there is a coupling in parsing the output data and we are not having the full picture of a remote job system. For now the new code chunk looks fairly isolated and can be taken out later when this local build service can be moved out as an external plugin and we have a better integration of remote job concepts.

grosser commented on app/controllers/builds_controller.rb:
```diff
@@ -92,8 +92,18 @@ def edit_build_params
-    DockerBuilderService.new(@build).run!(push: true)
+    docker_build_params = { push: true }
+    docker_build_params.update(extra_docker_build_params)
+    DockerBuilderService.new(@build).run!(docker_build_params)
```
this can be part of the plugin via a `build_params` hook

hdt91 replied:
What do you mean by ```build_params``` hook?

grosser replied:
nvm ... build_params would be attributes for the build ... this is something else ...

grosser commented on app/controllers/builds_controller.rb:
```diff
@@ -92,8 +92,18 @@ def edit_build_params
+  def extra_docker_build_params
+    hash = {}
+    Samson::Hooks.fire(:extra_build_params).each do |p|
+      hash[p] = params[p].present?
```
this sets true to some keys ... looks strange ... could merge extra_build_params ?

hdt91 replied:
This only does the minimum since for this task there is only one extra argument. If there are extra arguments the function can change to suit the need with extra conversions from string values to proper parameter values.

grosser replied:
atm it only supports booleans which is a bit strange ...

grosser replied:
idk ... good enough for now ...

grosser commented on app/models/docker_builder_service.rb:
```diff
@@ -32,6 +35,31 @@ def run!(image_name: nil, push: false, tag_as_latest: false)
     JobExecution.start_job(job_execution)
   end
 
+  def run_build_image_job(local_job, image_name, push: false, tag_as_latest: false)
```
can this code live somewhere else ? (not mix kubernetes into samson core)

hdt91 replied:
I would prefer to do this as well but as I said in the earlier comment, this piece of code has a coupling with ```DockerBuilderService``` in parsing the final output and to certain extent kinda of intertwine a local running job and a remote running job with logs content and job status passed around.

grosser commented on plugins/kubernetes/app/models/kubernetes/build_job_executor.rb:
```diff
@@ -0,0 +1,140 @@
+      k8s_job = Kubernetes::RoleConfigFile.new(contents, build_job_config_path).job
+      fill_job_details(k8s_job, build, project, tag: tag, push: push, tag_as_latest: tag_as_latest)
+
+      if k8s_job
```
if these is no job file this would all blow up ... now if needed ?

also these past 5-6 lines can be a method ?

hdt91 replied:
If there is no job file some succeeding code will likely trigger an exception and be caught in the ```ensure``` block. Is there anything wrong that I am not aware of?

What is your thought on grouping other lines of code into a method? These line look like an one-time glue code between the ```fill_job_details``` and ```create_and_wait_for_job``` to me.

grosser replied:
28-35 looks like one concern to me something like 'job_config'

grosser replied:
the File.read will already blow up, so I don't think we need a if k8s_job check ...

grosser commented on plugins/kubernetes/app/models/kubernetes/build_job_executor.rb:
```diff
@@ -0,0 +1,140 @@
+        @output.puts "### Remote build job #{job_name} completed sucessfully"
+      else
+        @output.puts "### Remote build job #{job_name} failed to run or timed out"
+      end
```
little dry-up with `message = (success ? "completed successfully" : "failed or timed out")`

grosser commented on plugins/kubernetes/app/models/kubernetes/build_job_executor.rb:
```diff
@@ -0,0 +1,140 @@
+    end
+
+    def execute!(build, project, tag:, push: false, registry:, tag_as_latest: false)
+      job_log = job_name = job_namespace = ""
```
will be nil anyway ... no need to assign ?

hdt91 replied:
Just my habit to declare all variables before using them to deal with code coverage or normal static analysis tool. Is this a bad thing in Ruby/Rails?

grosser replied:
nah not really just looks a bit strange ...

grosser commented on plugins/kubernetes/app/models/kubernetes/build_job_executor.rb:
```diff
@@ -0,0 +1,140 @@
+      if k8s_job
+        # Jobs will still be there regardless of their statuses
+        @output.puts "### Cleaning up the remote build job: #{job_name}"
+        extension_client.delete_job(job_name, job_namespace)
```
this wil blow up with a 404 when the job failed to create ?

hdt91 replied:
Yes, if it fails to create it will just throw 404 and be caught by the outer try/catch block of job executor, sending a proper message to the log. This ```ensure``` block is the best effort attempt to clean up resource remnants from the build job, what else do you think is good to add here?

grosser replied:
blowing up in the ensure will clobber the original error message, so I'd prefer a begin; delete_job; rescue KubeClient::Exception;end or so to silence it

grosser commented on plugins/kubernetes/app/models/kubernetes/build_job_executor.rb:
```diff
@@ -0,0 +1,140 @@
+      job_name = k8s_job[:metadata][:name]
+      job_namespace = k8s_job[:metadata][:namespace]
+
+      loop do
```
does a job have multiple pods ? ... not sure what's going on here ... some docs might help :)

hdt91 replied:
In this context I mean the build job should only have one container for simplicity. I do not know what kind of use case we need a build job with more than one container.

grosser replied:
not asking about containers, but why we need to loop here ... I was thinking we'd only have 1 pod created / don't need any looping logic

hdt91 replied:
Got it. I added a comment why it needs a loop. Sometimes the job is not yet ready and the pod log is not available so we need to retry a few seconds later.

grosser commented on app/views/builds/new.html.erb:
```diff
@@ -40,6 +40,8 @@
         </div>
       </div>
 
+      <%= Samson::Hooks.render_views(:build_new_use_kubernetes_job, self, form: form) %>
```
hook names should be universal, something like build_new for example

grosser commented on app/views/builds/show.html.erb:
```diff
@@ -61,6 +61,7 @@
   <% unless docker_build_running? @build %>
     <%= form_for :build, url: build_docker_image_project_build_path(@build.project, @build), html: { class: "form-inline" } do |form| %>
       <%= form.submit (job ? 'Rebuild Docker Image' : 'Build Docker Image'), class: "btn btn-primary" %>
+      <%= Samson::Hooks.render_views(:build_show_use_kubernetes_job, self, form: form) %>
```
same here -> build_show

grosser commented on plugins/kubernetes/test/models/kubernetes/build_job_executor_test.rb:
```diff
@@ -0,0 +1,152 @@
+require_relative "../../test_helper"
+
+SingleCov.covered! uncovered: 3
```
can you add a comment on why there are uncovered lines or possibly add some test to hit them ?

grosser commented on plugins/kubernetes/test/models/kubernetes/build_job_executor_test.rb:
```diff
@@ -0,0 +1,152 @@
+          assert_empty job_log
+        end
+
+        it 'job gets timeout' do
```
strange test case name ... how about something like `it does xyz when the job times out`

grosser commented on plugins/kubernetes/test/models/kubernetes/build_job_executor_test.rb:
```diff
@@ -0,0 +1,152 @@
+        assert_equal build_config.job[:spec][:template][:spec][:containers][0][:env].length, 1
+      end
+
+      describe 'can create job' do
```
describes usually are something about a certain state like "describe 'when all params are good' do ... it 'creates a job'"

grosser commented on plugins/kubernetes/test/test_helper.rb:
```diff
@@ -106,7 +106,11 @@ def kubernetes_fake_raw_template
     Kubernetes::ReleaseDoc.any_instance.stubs(raw_template: read_kubernetes_sample_file('kubernetes_deployment.yml'))
   end
 
+  def sample_file_path(file_name)
```
this is a global method so it needs to be namespaced to kubernetes ... kubernetes_sample_file_path

grosser commented on test/models/docker_builder_service_test.rb:
```diff
@@ -1,6 +1,6 @@
 require_relative '../test_helper'
 
-SingleCov.covered! uncovered: 19
+SingleCov.covered! uncovered: 20
```
what line did not get covered ... can we hit it somehow ?

hdt91 replied:
The ```run!``` method's body is uncovered and originally had 19 lines. I added an if check for running the remote job and now the body has one extra line. Do you have any idea how to test the code block passed to the job execution inside this method's body?

grosser replied:
nah ... just sanity checking whenever this increases ... you are welcome to
test it, but that can be a separate PR :)

On Wed, Jul 13, 2016 at 4:19 PM, Tran Hai Dang <notifications@github.com>
wrote:

> In test/models/docker_builder_service_test.rb
> <https://github.com/zendesk/samson/pull/1076#discussion_r70725527>:
>
> > @@ -1,6 +1,6 @@
> >  require_relative '../test_helper'
> >
> > -SingleCov.covered! uncovered: 19
> > +SingleCov.covered! uncovered: 20
>
> The run! method's body is uncovered and originally had 19 lines. I added
> an if check for running the remote job and now the body has one extra line.
> Do you have any idea how to test the code block passed to the job execution
> inside this method's body?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendesk/samson/pull/1076/files/3dca884387016cc9f2b1ffd3594c6cfbe9d9a34b#r70725527>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAAsZwX8mCKykD3mFr1AtS1C-6hu2i5Gks5qVXKHgaJpZM4I2Rcw>
> .
>

grosser commented on test/models/docker_builder_service_test.rb:
```diff
@@ -23,6 +23,33 @@
+      Kubernetes::BuildJobExecutor.expects(:new).returns k8s_job
+    end
+
+    it 'calls the job executor and the job completes successfully' do
```
something more descriptive like `it "does xyz when the job completes"`

grosser commented on test/models/docker_builder_service_test.rb:
```diff
@@ -23,6 +23,33 @@
+      assert_equal("#{project.docker_repo}@#{repo_digest}", build.docker_repo_digest)
+    end
+
+    it 'calls the job executor but it is not successful' do
```
it 'does xyz when the job fails'

hdt91 commented:
I updated this PR with better test case description and other minor cleanups. Still there is a red flag in CodeClimate, do you mind checking what is the issue there?

grosser commented:
"Analysis results are provided for code written by licensed individuals, and Hai Dang Tran does not appear to be a member of your organization ." ... so no idea ... maybe it's just not doing the report and that's why it's red ...

grosser commented on plugins/kubernetes/app/models/kubernetes/build_job_executor.rb:
```diff
@@ -0,0 +1,144 @@
+    private
+
+    def build_cluster
+      Kubernetes::Cluster.first!
```
this most likely needs some tweeking ... something like picking a context / cluster via env vars ... but good enough for now ...

grosser commented on plugins/kubernetes/app/views/samson_kubernetes/_build_show.erb:
```diff
@@ -0,0 +1,4 @@
+<div class="form-group">
+  <%= check_box_tag :kubernetes_job%>
+  <%= form.label :kubernetes_job, "Use Remote Kubernetes Job", class: "control-label" %>
```
should this be a column instead ? ... then we don't need extra_build_params ... just build_params and let it persist ... and no additional UI for build/show ...

hdt91 replied:
If this is persisted as a build's parameter, it becomes a part of core Samson workflow, doesn't it?

grosser replied:
nah ... just an extra column (default false) nobody needs to know about :)

hdt91 replied:
Sounds good to me. Only one small issue is to switch between local build and remote build option without an UI. What is your suggestion?

grosser replied:
the new build form gets checkbox ... rebuild does not really need it ... can always make a new build if someone wants to build without remote

hdt91 replied:
Updated with the new column :)

grosser commented:
looking good ❤️

grosser commented:
looks good to me ... @jonmoter / @rata want to give this a sanity check before I merge ?

jonmoter commented:
Just did a quick review.  Nothing jumped out to me as a blocker.  Let's merge, try it out for a bit, and iterate if need be.

👍

hdt91 commented:
Thanks for very quick feedback guys :D

rata commented on app/models/docker_builder_service.rb:
```diff
@@ -91,6 +119,20 @@ def project
+    {
+      username: ENV['DOCKER_REGISTRY_USER'],
+      password: ENV['DOCKER_REGISTRY_PASS'],
+      email: ENV['DOCKER_REGISTRY_EMAIL'],
```
email is deprecated and will be removed in docker 1.13 IIRC. Maybe we can remove it now?

rata commented on plugins/kubernetes/app/models/kubernetes/build_job_executor.rb:
```diff
@@ -0,0 +1,144 @@
+      # Fill in some information to easily query the job resource
+      project_name = project.permalink.tr('_', '-')
+      labels = { project: project_name, role: "docker-build-job" }
+      k8s_job[:metadata][:name] = "#{project_name}-docker-build-#{build.id}-#{SecureRandom.hex(7)}"
```
why not generateName, instead of name, that does this?

rata commented:
@grosser thanks, seems fine!
